File: 2024/01/code.cpp
#include <iostream>
#include <fstream>

#include <sstream>
#include <cmath>
#include <map>

using namespace std;

vector<int> list_1;
vector<int> list_2;
map<int, int> map_2;

vector<string> splitString(string input) {
    vector<string> tokens;
    stringstream stream(input);
    string token;

    while (stream >> token) {
        tokens.push_back(token);
    }

    return tokens;
}

void parseFile(string inputFile) {
    string line;
    ifstream f(inputFile);
    while (getline(f, line)) {
        vector<string> nums = splitString(line);

        list_1.push_back(stoi(nums[0]));
        list_2.push_back(stoi(nums[1]));

        map_2[stoi(nums[1])] += 1;
    }

    f.close();
}

int part1() {
    sort(list_1.begin(), list_1.end());
    sort(list_2.begin(), list_2.end());

    int sum = 0;
    for (size_t i = 0; i < list_1.size(); i++) {
        sum += abs(list_1[i] - list_2[i]);
    }

    return sum;
}

int part2() {
    int sum = 0;
    for (size_t i = 0; i < list_1.size(); i++) {
        sum += list_1[i] * map_2[list_1[i]];
    }

    return sum;
}

int main(int argc, char** argv) {
    if (argc != 2) {
        cout << "Wrong number of arguments: " << argc << "\n";
        return 1;
    }
    string inputFile = argv[1];

    parseFile(inputFile);
    cout << "Part 1 answer: " << part1() << "\n";
    cout << "Part 2 answer: " << part2() << "\n";
    return 0;
}

File: 2024/01/code.py
import sys

from collections import Counter

list_1 = []
list_2 = []

def parse_file(input_path: str):
    with open(input_path) as f:
        for line in f:
            num_1, num_2 = line.split()
            list_1.append(int(num_1))
            list_2.append(int(num_2))

def part_1():
    sum = 0

    list_1.sort()
    list_2.sort()
    for a, b in zip(list_1, list_2):
        sum += abs(a - b)

    print(f"Part 1 answer: {sum}")

def part_2():
    sum = 0

    counter_2 = Counter(list_2)
    for num in list_1:
        sum += num * counter_2.get(num, 0)

    print(f"Part 2 answer: {sum}")

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(f"Wrong number of arguments: {len(sys.argv)}")
        sys.exit(1)
    input_path = sys.argv[1]

    parse_file(input_path)
    part_1()
    part_2()
